### ISSUE TITLE
Handle closed twice when OpenVPNClient destroyed and using wintun

### ISSUE THREAD

**Issue by AnthonyWharton** [OpenVPN/openvpn3#110]
I have a class that inherits `::openvpn::ClientAPI::OpenVPNClient` that works perfectly with the tap driver, however when experimenting with the wintun support I've noticed an uncaught exception when destroying the class: 

```
Exception thrown at 0x00007ffc90dfff0a (ntdll.dll) in test-client.exe: 0xC0000008: An invalid handle was specified.
```

When the `openvpn::TunWin::WintunClient`'s, `ring_buffer`'s, `send_ring_tail_moved_` field is destroyed, the closing of the handle within the `Win::Event` object throws the exception.  This would suggest the handle has already been closed, but I am unable to find where this might be.

One possible lead that I see is that the `send_ring_tail_moved_` field is assigned to the `AsioEvent`, `send_tail_moved_asio_event_` during the `io_context` overload of the `RingBuffer`'s constructor.  The highly asyncronous nature of asio code makes it quite hard to follow, and as I am not super familiar with it lost the scent of a potential lead there.

If you have any ideas what could be causing this issue I would be very grateful!

The full stack trace is below:

```
ntdll.dll!00007ffc90dfff0a() (Unknown Source:0)
KernelBase.dll!00007ffc8df04a35() (Unknown Source:0)
openvpn::Win::ScopedHANDLE::close() Line 90 (<openvpn-src>\openvpn\win\scoped_handle.hpp:90)
openvpn::Win::ScopedHANDLE::~ScopedHANDLE() Line 101 (<openvpn-src>\openvpn\win\scoped_handle.hpp:101)
openvpn::Win::Event::~Event() (Unknown Source:0)
openvpn::TunWin::RingBuffer::~RingBuffer() Line 124 (<openvpn-src>\openvpn\tun\win\ringbuffer.hpp:124)
openvpn::TunWin::RingBuffer::`scalar deleting destructor'(unsigned int) (Unknown Source:0)
openvpn::intrusive_ptr_release<openvpn::TunWin::RingBuffer>(openvpn::TunWin::RingBuffer * p) Line 721 (<openvpn-src>\openvpn\common\rc.hpp:721)
openvpn::RCPtr<openvpn::TunWin::RingBuffer>::~RCPtr<openvpn::TunWin::RingBuffer>() Line 140 (<openvpn-src>\openvpn\common\rc.hpp:140)
openvpn::TunWin::WintunClient::~WintunClient() (Unknown Source:0)
openvpn::TunWin::WintunClient::`vbase destructor'() (Unknown Source:0)
openvpn::TunWin::WintunClient::`scalar deleting destructor'(unsigned int) (Unknown Source:0)
openvpn::intrusive_ptr_release<openvpn::TunClient>(openvpn::TunClient * p) Line 721 (<openvpn-src>\openvpn\common\rc.hpp:721)
openvpn::RCPtr<openvpn::TunClient>::~RCPtr<openvpn::TunClient>() Line 140 (<openvpn-src>\openvpn\common\rc.hpp:140)
openvpn::ClientProto::Session::~Session() Line 265 (<openvpn-src>\openvpn\client\cliproto.hpp:265)
openvpn::ClientProto::Session::`scalar deleting destructor'(unsigned int) (Unknown Source:0)
openvpn::intrusive_ptr_release<openvpn::ClientProto::Session>(openvpn::ClientProto::Session * p) Line 721 (<openvpn-src>\openvpn\common\rc.hpp:721)
openvpn::RCPtr<openvpn::ClientProto::Session>::~RCPtr<openvpn::ClientProto::Session>() Line 140 (<openvpn-src>\openvpn\common\rc.hpp:140)
openvpn::ClientConnect::~ClientConnect() Line 285 (<openvpn-src>\openvpn\client\cliconnect.hpp:285)
openvpn::ClientConnect::`scalar deleting destructor'(unsigned int) (Unknown Source:0)
openvpn::intrusive_ptr_release<openvpn::ClientConnect>(openvpn::ClientConnect * p) Line 721 (<openvpn-src>\openvpn\common\rc.hpp:721)
openvpn::RCPtr<openvpn::ClientConnect>::~RCPtr<openvpn::ClientConnect>() Line 140 (<openvpn-src>\openvpn\common\rc.hpp:140)
openvpn::RCPtr<openvpn::ClientConnect>::reset() Line 157 (<openvpn-src>\openvpn\common\rc.hpp:157)
openvpn::ClientAPI::Private::ClientState::~ClientState() Line 518 (<openvpn-src>\client\ovpncli.cpp:518)
openvpn::ClientAPI::Private::ClientState::`scalar deleting destructor'(unsigned int) (Unknown Source:0)
openvpn::ClientAPI::OpenVPNClient::~OpenVPNClient() Line 1421 (<openvpn-src>\client\ovpncli.cpp:1421)
MyClient::~MyClient() (Unknown Source:0)
<calls to reset() a MyClient wrapped in a std::unique_ptr>
```

**Comment by lstipakov** [comment]
IIRC `CloseHandle()` throws an exception only when running under debugger.

Briefly looked at the code and your explanation makes sense - indeed handle is shared between `send_tail_moved_asio_event_` (instance of `AsioEvent`) and `send_ring_tail_moved_`(instance of `Win::Event`). It appear that first `AsioEvent` closes handle and then `WinEvent` tries to close it again.

**Comment by AnthonyWharton** [comment]
Thank you kindly @lstipakov!

I’m still not too proficient with lower level Win32 development and didn’t realise it would be as simple as just duplicating the handle! Thanks for that, I’ll apply that patch locally and try again now.

**Comment by AnthonyWharton** [comment]
Tested duplicating the handle and that seems to be stable.

Left the proposed fix in PR #111 for future merging. Thanks for your help.
